### PR TITLE
Created Approaching-a-DSA-problem.md

### PR DESCRIPTION
**Pull request by triach-rold**
## Description

Wroteup a quick "starters guide" to approaching a DSA problem in a logical way
Fixes #(Insert Issue Here Without Brackets)

## Changes Include

File created. (Approaching-a-DSA-problem.md)

### CODE REVIEW

File: src/faqs/approaching-a-DSA-problem.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# How to approach a DSA problem ?
+#### Understanding the problem.
+The first step in approaching a DSA problem is to understand the essence of the problem itself. We must try and figure out what the problem demands. Not understanding the basic working of the problem is a big mistake. If the wording of the question is unclear, look at the testcases given for the particular problem and try to reverse-engineer the demands of the question.
```
Comment (Bhupesh-V) [inline src/faqs/approaching-a-DSA-problem.md:3]:
```suggestion
- The first step in approaching a DSA problem is to understand the essence of the problem itself. We must try and figure out what the problem demands. 
- Not understanding the basic needs of the problem asked is a big mistake. If the wording of the question is unclear, look at the testcases given for the particular problem and try to reverse-engineer the demands of the question.
```

File: src/faqs/approaching-a-DSA-problem.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+#### Clearing out a logical path.
+Now that the problem is understood, we must move on to the logic. We must have a clear picture of the answer's logic in our head. Once we have a clear "map" of this logic, we move on to implementing it in code. If the logic of your answer statement is fairly complicated, consider writing down comments while implementing the logic in your code. Once the logic is thought of, try to think of alternate ways to approach the problem if it doesn't meet the complexity constraints. Repeat this step until you're happy with your desired logic for the answer. Using pseudocode or diagrammatic representations to refine the algorithm before coding can help to understand the logic and essence of the program well.
```
Comment (Bhupesh-V) [inline src/faqs/approaching-a-DSA-problem.md:5]: We can reword the first point, to avoid repetition of the sentiment "need to build a clear picture"

```suggestion
- The 2nd step involves building a clear picture of the answer's logic in our head. If the logic of your answer statement is fairly complicated, consider writing down comments while implementing the logic in your code. 
- Once the logic is thought of, try to think of alternate ways to approach the problem if it doesn't meet the complexity constraints. Repeat this step until you're happy with your desired logic for the answer. Using pseudocode or diagrammatic representations to refine the algorithm before coding can help to understand the logic and essence of the program well.
```

File: src/faqs/approaching-a-DSA-problem.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+#### Effective Implementation. 
+Implement the logic in your code. As said before, consider adding comments to your code to make the logic "flow" better while solving it. Eventually, with enough problem solving, you won't require the comments to implement logic(s) anymore, though it is a good habit to do so. Another "cosmetic" tip is to name your variables well. Try to use as little variables as possible to minimize the storage space required for your code. Try to include various problem-solving paradigms such as greedy, dynamic programming, divide and conquer paradigms to guide the design of your algorithms.
```
Comment (Bhupesh-V) [inline src/faqs/approaching-a-DSA-problem.md:7]:
```suggestion
- Implement the logic in your code. As said before, consider adding comments to your code to make the logic "flow" better while solving it. Eventually, with enough problem solving, you won't require the comments to implement logic(s) anymore, though it is a good habit to do so. 
- Another "cosmetic" tip is to name your variables well. Try to use as little variables as possible to minimize the storage space required for your code. Try to include various problem-solving paradigms such as greedy, dynamic programming, divide and conquer paradigms to guide the design of your algorithms.
```

File: src/faqs/approaching-a-DSA-problem.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+#### Testing and submission.
+Once implemented, test. Your code should be tested on a exhaustive set of inputs and outputs. This exhaustive search will allow you to find and fix any edge cases that slipped under the radar the first time. Try to think about why your code breaks at that edge case and see if it's a basic logical fallacy in your code, or simply a case that could have slipped between the cracks during the implementation. Try  using systematic debugging techniques such as print statements, debugging tools, or test case generators to identify and resolve errors more efficiently. This would allow you to pinpoint the source of all the errors and allow you to fix it surgically.
```
Comment (Bhupesh-V) [inline src/faqs/approaching-a-DSA-problem.md:9]:
```suggestion
- Once implemented, test. Your code should be tested on a exhaustive set of inputs and outputs. This exhaustive testing will allow you to find and fix any edge cases that slipped under the radar the first time. 
- Try to think about why your code breaks at that edge case and see if it's a basic logical fallacy in your code, or simply a case that could have slipped between the cracks during the implementation. 
- Try  using systematic debugging techniques such as print statements, debugging tools, or test case generators to identify and resolve errors more efficiently. This would allow you to pinpoint the source of all the errors and allow you to fix it surgically.
```

File: src/faqs/approaching-a-DSA-problem.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+#### Dealing with fallacies.
+If all else has failed, try a brute force method to meet all the constraints of the problem and work your way from there. Try to look at your logic and think about ways it can be optimized, even at the cost of storage space. Try to think of any way to make your code run faster. Try to think of more "elegant" ways that would not require recursion in any form. Thinking about creative ways to solve problems will allow you to level up your skill as well as make you think about the logic of your program more deeply.
```
Comment (Bhupesh-V) [inline src/faqs/approaching-a-DSA-problem.md:11]: Loved this section!

```suggestion
- If all else has failed, try a brute force method to meet all the constraints of the problem and work your way from there. Try to look at your logic and think about ways it can be optimized, even at the cost of storage space. Try to think of any way to make your code run faster. 
- Try to think of more "elegant" ways that would not require recursion in any form. Thinking about creative ways to solve problems will allow you to level up your skill as well as make you think about the logic of your program more deeply.
```

File: src/faqs/approaching-a-DSA-problem.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+#### Effective debugging.
+Breaking your code down into smaller parts would allow you to approach the problem in a more appealing way, allowing you to debug and test it easily. Identify patterns between logic(s) of various problems and think about why they resemble each other. Try thinking about possible tradeoffs between speed and storage. Balancing both is key in real world situations.
```
Comment (Bhupesh-V) [inline src/faqs/approaching-a-DSA-problem.md:13]: I guess the intention here was to talk about "memory (RAM usage)" than "storage (Disk storage)" right?

```suggestion
- Breaking your code down into smaller parts would allow you to approach the problem in a more appealing way, allowing you to debug and test it easily. Identify patterns between logic(s) of various problems and think about why they resemble each other. 
- Try thinking about possible tradeoffs between **speed** and **memory usage**. Balancing both is key in real world situations. 
```

File: src/faqs/approaching-a-DSA-problem.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+#### Effective debugging.
+Breaking your code down into smaller parts would allow you to approach the problem in a more appealing way, allowing you to debug and test it easily. Identify patterns between logic(s) of various problems and think about why they resemble each other. Try thinking about possible tradeoffs between speed and storage. Balancing both is key in real world situations. 
+#### Patience is key.
+Most importantly, you must be patient throughout the whole process. Getting stuck is a part of the journey. Apart from that, solving more DSA problems will make you better at it. That is pretty much true for every replicable activity. Solving more problems, especially harder questions would allow you to step out of your comfort zone by a lot. Learning new algorithms, mathematical formulae, etc are all things that would make you a better programmer in the long run.
```
Comment (Bhupesh-V) [inline src/faqs/approaching-a-DSA-problem.md:15]:
```suggestion
- Most importantly, you must be patient throughout the whole process. Getting stuck is a part of the journey. Apart from that, solving more DSA problems will make you better at it. That is pretty much true for every replicable activity. Solving more problems, especially harder questions, would allow you to step out of your comfort zone by a lot. 
- Learning new algorithms, mathematical formulae, etc. are all things that would make you a better programmer in the long run. 
```

Comment (Bhupesh-V) [inline src/faqs/approaching-a-DSA-problem.md:15]: Would be nice if you would like to link any of your public profiles, we want to make sure folks who put effort writing wiki need to be acknowledged 💙


```suggestion
Most importantly, you must be patient throughout the whole process. Getting stuck is a part of the journey. Apart from that, solving more DSA problems will make you better at it. That is pretty much true for every replicable activity. Solving more problems, especially harder questions would allow you to step out of your comfort zone by a lot. Learning new algorithms, mathematical formulae, etc are all things that would make you a better programmer in the long run. 

## Attribution

This wiki was maintained/written by [PROFILE](LINK)
```

**Review by Bhupesh-V — CHANGES_REQUESTED** [review]
Thanks a lot for writing this, I think it captures almost everything for this topic. Having said that, I have provided suggestions to divide paragraphs into points (with very minor modifications) so that they are much easier to read.

Also, if possible, I would recommend adding any additional "meta (approaching DSA/coding problems)" reading material in the end.

**Comment by Bhupesh-V** [comment]
@triach-rold One last thing, please also update the [SUMMARY file](https://github.com/developersIndia/wiki/blob/main/src/SUMMARY.md) so that it's visible on the Wiki

File: src/faqs/approaching-a-DSA-problem.md (hunk 6 above)
Comment (triach-rold) [reply src/faqs/approaching-a-DSA-problem.md]: It should be RAM, yes. Thanks for pointing out the errata.

**Review by Bhupesh-V — APPROVED** [review]
Thanks for your contribution @triach-rold   🧡

We will be announcing this wiki across our channels, in the coming week. If you want us to tag you anywhere (Twitter, LinkedIn, Reddit, Discord), please email us your profiles here: contact@developersindia.in

**Comment by Bhupesh-V** [comment]
@all-contributors please add @triach-rold for content
